### PR TITLE
feat: Add synchronization to start of audio recordings

### PR DESCRIPTION
**Pull request by Om1609**
## Summary

When a Voice Channel is being recorded, users' recordings start when they start speaking, and not when the recording was actually started. This PR adds silence at the start of the recordings of users who weren't speaking when the recording started, but started speaking later.

This however relies on the receipt timestamp of the first packet, which can sometimes cause slightly inaccurate synchronisation, based on the network conditions.

This PR solves #1980.

## Information

- [x] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed).
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting,
      examples, ...).

## Checklist

- [x] I have searched the open pull requests for duplicates.
- [x] If code changes were made then they have been tested.
  - [x] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why.
- [x] I have updated the changelog to include these changes.

### CODE REVIEW

**Comment by Om1609** [comment]
I'll add an example for this too. For now, this is enough to test-
```py
import asyncio
import io

from pydub import AudioSegment

import discord
from discord.ext import commands
from discord.sinks import MP3Sink


class MyCog(commands.Cog):
    def __init__(self, bot: commands.Bot):
        self.bot = bot

    async def finished_callback(
        self,
        sink: MP3Sink,
        channel: discord.TextChannel,
    ):
        mention_strs = []
        audio_segs: list[AudioSegment] = []
        files: list[discord.File] = []

        longest = AudioSegment.empty()

        for user_id, audio in sink.audio_data.items():

            mention_strs.append(f"<@{user_id}>")

            seg = AudioSegment.from_file(audio.file, format="mp3")

            # Determine the longest audio segment
            if len(seg) > len(longest):
                audio_segs.append(longest)
                longest = seg
            else:
                audio_segs.append(seg)

            audio.file.seek(0)
            files.append(discord.File(audio.file, filename=f"{user_id}.mp3"))

        for seg in audio_segs:
            longest = longest.overlay(seg)

        with io.BytesIO() as f:
            longest.export(f, format="mp3")
            await channel.send(
                f"Finished! Recorded audio for {', '.join(mention_strs)}.",
                files=files + [discord.File(f, filename="recording.mp3")],
            )

    @commands.command()
    async def join(self, ctx: commands.Context):
        await ctx.author.voice.channel.connect()

    @commands.command()
    async def record(self, ctx: commands.Context):
        vc: discord.VoiceClient = ctx.voice_client

        await ctx.send("Recording...")

        vc.start_recording(
            discord.sinks.MP3Sink(),
            self.finished_callback,
            ctx.channel,
            sync_start=True,
        )

        await asyncio.sleep(15) # edit to change recording duration
        vc.stop_recording()
        await ctx.send("Stopped recording.")


def setup(bot: commands.Bot):
    bot.add_cog(MyCog(bot))
```

**Comment by Lulalaby** [comment]
Sigh

File: discord/voice_client.py [hunk 1]
```diff
@@ -700,7 +700,7 @@ def unpack_audio(self, data):
 
         self.decoder.decode(data)
 
-    def start_recording(self, sink, callback, *args):
+    def start_recording(self, sink, callback, *args, sync_start=False):
```
Comment (VincentRPS) [inline discord/voice_client.py:703]: little nitpick
```suggestion
    def start_recording(self, sink, callback, *args, sync_start: bool = False):
```

**Review by VincentRPS — DISMISSED** [review]

**Comment by Lulalaby** [comment]
I'm gonna review that internally with my own team. Until then on hold.

**Comment by JustaSqu1d** [comment]
Funny how this PR is 1984...

File: examples/audio_recording_merged.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
+import io
+
+import pydub  # pip install pydub
```
Comment (JustaSqu1d) [inline examples/audio_recording_merged.py:3]: Maybe add a version here. Such as 
```suggestion
import pydub  # pip install pydub==0.25.1
```

**Review by JustaSqu1d — CHANGES_REQUESTED** [review]

File: examples/audio_recording_merged.py (hunk 2 above)
Comment (Om1609) [reply examples/audio_recording_merged.py:3]: Sure

**Comment by plun1331** [comment]
@Lulalaby any updates?

**Comment by Om1609** [comment]
Check [this message](https://discord.com/channels/881207955029110855/881735314987708456/1097582070903095387) by Lala in the discord server

**Comment by BobDotCom** [comment]
Please fix merge conflicts

**Review by Lulalaby — DISMISSED** [review]

**Comment by Lulalaby** [comment]
i rly had a stroke rn lol

**Comment by Lulalaby** [comment]
@plun1331 ping pong

**Review by plun1331 — APPROVED** [review]

**Review by Lulalaby — APPROVED** [review]

**Pull request merged**